conjure-macros/src/endpoints.rs

```diff
@@ -0,0 +1,705 @@
+    let decoder = arg.params.decoder.as_ref().map_or_else(
+        || quote!(conjure_http::server::FromStrDecoder),
+        |d| quote!(#d),
+    );
+    quote! {
+        let #name = conjure_http::private::query_param::<_, #decoder>(&#query_params, #key, #param)?;
+    }
+}
+
+fn generate_header_arg(parts: &TokenStream, arg: &Arg<ParamArg>) -> TokenStream {
+    let name = &arg.ident;
+    let header = &arg.params.name;
+    let param = arg.ident.to_string();
+    let decoder = arg.params.decoder.as_ref().map_or_else(
+        || quote!(conjure_http::server::FromStrDecoder),
+        |d| quote!(#d),
+    );
+    quote! {
+        let #name = conjure_http::private::header_param::<_, #decoder>(&#parts, #header, #param)?;
+    }
+}
+
+fn generate_auth_arg(parts: &TokenStream, arg: &Arg<AuthArg>) -> TokenStream {
+    let name = &arg.ident;
+    let call = match &arg.params.cookie_name {
+        Some(cookie_name) => {
+            let prefix = format!("{}=", cookie_name.value());
+            quote!(parse_cookie_auth(&#parts, #prefix))
+        }
+        None => quote!(parse_header_auth(&#parts)),
+    };
+
+    quote! {
+        let #name = conjure_http::private::#call?;
+    }
+}
+
+fn generate_body_arg(
+    parts: &TokenStream,
+    body: &TokenStream,
+    service: &Service,
+    arg: &Arg<BodyArg>,
+) -> TokenStream {
+    let name = &arg.ident;
+    let trait_ = match service.asyncness {
+        Asyncness::Sync => quote!(DeserializeRequest),
+        Asyncness::Async => quote!(AsyncDeserializeRequest),
+    };
+    let deserializer = arg.params.deserializer.as_ref().map_or_else(
+        || quote!(conjure_http::server::ConjureRequestDeserializer),
+        |d| quote!(#d),
+    );
+    let await_ = match service.asyncness {
+        Asyncness::Sync => quote!(),
+        Asyncness::Async => quote!(.await),
+    };
+    quote! {
+        let #name = <#deserializer as conjure_http::server::#trait_<_, _>>::deserialize(
+            &#parts.headers,
+            #body,
+        ) #await_ ?;
+    }
+}
+
+fn generate_context_arg(
+    parts: &TokenStream,
+    response_extensions: &TokenStream,
+    arg: &Arg<ContextArg>,
+) -> TokenStream {
+    let name = &arg.ident;
+    quote! {
+        let #name = conjure_http::server::RequestContext::new2(&#parts, #response_extensions);
+    }
+}
+
+fn generate_response(
+    parts: &TokenStream,
+    response: &TokenStream,
+    service: &Service,
+    endpoint: &Endpoint,
+) -> TokenStream {
+    let trait_ = match service.asyncness {
+        Asyncness::Sync => quote!(SerializeResponse),
+        Asyncness::Async => quote!(AsyncSerializeResponse),
+    };
+    let serializer = endpoint.params.produces.as_ref().map_or_else(
+        || quote!(conjure_http::server::EmptyResponseSerializer),
+        |s| quote!(#s),
+    );
+
+    quote! {
+        <#serializer as conjure_http::server::#trait_<_, _>>::serialize(
+            &#parts.headers,
+            #response,
+        )
+    }
+}
+
+struct Service {
```

Comment:
This setup is moving towards a world in which we can use the same logic for custom and normal Conjure service generation re #245